### ISSUE TITLE
Bracket Pair Colorization for HTML tags

### ISSUE THREAD

**Issue by tgrelka** [microsoft/vscode#137504]
I love the new, native built-in bracket pair colorization feature of VS Code, colorful brackets are useful for quickly identifying the current block.

However, I still miss colorization in HTML files and templates. There are some extensions, but these have the same shortcomings as the (now replaced) Rainbow Brackets 2 extension, as well as quite “ugly” rendering.
IntelliJ IDEA has a popular plugin (called Rainbow Brackets as well) which handles HTML (and XML) opening and closing tags very nicely:
The tag name itself still has the default highlighting, in my case it’s a light green. However, the angle brackets `<>` and forward slash `/` are colorized according to the block level.

Please take a look at the attached example of this:
![image](https://user-images.githubusercontent.com/22134144/142617723-b42c2280-9aed-48d9-81eb-f3d39a1c75e1.png)

It would be really nice to have a similar feature with similar rendering in VS Code itself, because especially for structured files like HTML, with lots of nesting, more color would be beneficial for a quick and at-a-glance overview of the document structure.

**Comment by vintprox** [comment]
Currently, there is a workaround in VS Code:
```
  "editor.language.colorizedBracketPairs": [
    ["[", "]"],
    ["(", ")"],
    ["{", "}"],
    ["<", "</"],
    ["<", "/>"]
  ],
```

But it's buggy. It doesn't understand language specifics for identifying something as "bracket". It can recognize a bracket on something that is supposed to be a self-closing tag and lead to incorrect pair, for example.

**Comment by Sequoya42** [comment]
Thanks @vintprox 
Its indeed buggy and does not match correctly for `< />`  (like img tag).
Would love for this to be integrated in vscode, would make reading HTML 10x more confortable

**Comment by SpacemanOG** [comment]
I just configured my system following the instructions given here: https://www.stefanjudis.com/snippets/how-to-configure-vs-code-bracket-colors/. However, it doesn't work for HTML! Any suggestions on whether there's a way to solve it?

**Comment by MuhsinEmreAYGAR** [comment]
> I just configured my system following the instructions given here: https://www.stefanjudis.com/snippets/how-to-configure-vs-code-bracket-colors/. However, it doesn't work for HTML! Any suggestions on whether there's a way to solve it?

I had this problem too. Did you solve?

**Comment by OtherHorizon** [comment]
> I just configured my system following the instructions given here: https://www.stefanjudis.com/snippets/how-to-configure-vs-code-bracket-colors/. However, it doesn't work for HTML! Any suggestions on whether there's a way to solve it?
Referring [this](https://github.com/microsoft/vscode/issues/159586#issuecomment-1256404578) does it help?

**Comment by SpacemanOG** [comment]
Mine also doesn’t work for HTML

Sent from my iPhone

On Sep 25, 2022, at 00:58, Blip blop ***@***.***> wrote:

﻿

I just configured my system following the instructions given here: https://www.stefanjudis.com/snippets/how-to-configure-vs-code-bracket-colors/. However, it doesn't work for HTML! Any suggestions on whether there's a way to solve it?
Referring this<https://github.com/microsoft/vscode/issues/159586#issuecomment-1256404578> does it help?

**Comment by daslicht** [comment]
+1 This would be a very useful feature !

**Comment by dnhuan** [comment]
Looking forward to this feature!

**Comment by juanchavezlive** [comment]
Bump !

**Comment by nano0018** [comment]
Bump!

**Comment by daslicht** [comment]
A connecting line woudl be very usefull either 
eg like this but for html tags:
![](https://i.imgur.com/lkHK0ZU.png)

**Comment by alan-venv** [comment]
Bump !

**Comment by kemoboy** [comment]
bump

**Comment by davyd-souza** [comment]
+1

**Comment by Sov3rain** [comment]
Would be very nice, +1!

**Comment by Ken-SBIT** [comment]
This would be a most-appreciated addition! +1

**Comment by brunomaria71** [comment]
need this!

**Comment by AlexanderGeorgiou** [comment]
Bumppp

**Comment by alejandro-venegas** [comment]
Bump

**Comment by starball5** [comment]
Related on Stack Overflow: [VSCode - Bracket Pair Colorization in HTML/CSS (How to?)](https://stackoverflow.com/q/73358182/11107541)

**Comment by dawidmachon** [comment]
bump

**Comment by Sdju** [comment]
bump

**Comment by developer-256** [comment]
You don't need opening and closing tags to be colored. What you need is the Indent Guides to be colored, and you can achieve this without any extension.

<img width="202" alt="code" src="https://github.com/microsoft/vscode/assets/121548329/9fd963c6-b041-48f5-ba23-f5311745ed29">


1. Hit **Ctrl+P** in VS Code.
2. Write **>settings.json** and hit Enter.
3. Paste the following code:

`"workbench.colorCustomizations": {
        "editorIndentGuide.activeBackground1": "#c3e88d"
    },`

That should be it. 

**Change the color of indent guide:**
You can also change the color of indent guide to whatever you like by simply modifying the hex value to your desired color or by hovering over the color and selecting the one you prefer.

`"workbench.colorCustomizations": {
        "editorIndentGuide.activeBackground1": "#ffcb61"
    },`

<img width="199" alt="code2" src="https://github.com/microsoft/vscode/assets/121548329/625a85e5-4e67-4b30-9cd2-57a3fb8711b9">

_Also this works with every other programming language._

**Comment by RudyRSDev** [comment]
anyone still on the search for this

**Comment by daslicht** [comment]
> You don't need opening and closing tags to be colored. What you need is the Indent Guides to be colored, and you can achieve this without any extension.
> 
> <img alt="code" width="202" src="https://user-images.githubusercontent.com/121548329/281112800-9fd963c6-b041-48f5-ba23-f5311745ed29.png">
> 1. Hit **Ctrl+P** in VS Code.
> 2. Write **>settings.json** and hit Enter.
> 3. Paste the following code:
> 
> `"workbench.colorCustomizations": { "editorIndentGuide.activeBackground1": "#c3e88d" },`
> 
> That should be it.
> 
> **Change the color of indent guide:** You can also change the color of indent guide to whatever you like by simply modifying the hex value to your desired color or by hovering over the color and selecting the one you prefer.
> 
> `"workbench.colorCustomizations": { "editorIndentGuide.activeBackground1": "#ffcb61" },`
> 
> <img alt="code2" width="199" src="https://user-images.githubusercontent.com/121548329/281113880-625a85e5-4e67-4b30-9cd2-57a3fb8711b9.png">
> _Also this works with every other programming language._

doesnt work here:
![](https://i.imgur.com/4mKP2kF.png)

**Comment by developer-256** [comment]
> > That should be it.
> 
> doesnt work here, no matter where i click the tag: ![](https://camo.githubusercontent.com/a37771682f29fa9441364c38ee34071092c787e74796dc403392f806886a5256/68747470733a2f2f692e696d6775722e636f6d2f346d4b50326b462e706e67)

I have tried it in php and it is working quite fine.

Just uninstall **Toggle Bracket Guides** extension and also delete `"editor.guides.bracketPairs": true,` command in settings.json and then add 
```"workbench.colorCustomizations": {
        "editorIndentGuide.activeBackground1": "#ffcb61"
    },```
or add `"editorIndentGuide.activeBackground1": "#ffcb61"` within **{}** of **"workbench.colorCustomizations": {}** if it is present by default. 

For reference here is the php code screenshot.
<img width="235" alt="image" src="https://github.com/microsoft/vscode/assets/121548329/9bc72f45-db7a-4315-9ec5-9ac3c2cf653d">

You can clearly see yellow indent line

**Comment by developer-256** [comment]
> > That should be it.
> 
> doesnt work here, no matter where i click the tag


It works where code is indented and also uninstall Toggle Bracket Guides extension and remove `"editor.guides.bracketPairs": true,` for settings.json as it cancels the effect of above command

**Comment by developer-256** [comment]
> > That should be it.
> 
> doesnt work here, no matter where i click the tag

<img width="87" alt="image" src="https://github.com/microsoft/vscode/assets/121548329/51b56643-53d3-41c0-ac03-1952190f5d89">

**Comment by henrikvilhelmberglund** [comment]
> 
> It works where code is indented and also uninstall Toggle Bracket Guides extension and remove `"editor.guides.bracketPairs": true,` for settings.json as it cancels the effect of above command

I don't want to disable bracketPairs though because I like it for coding. The whole point of bracket pair colorization is that the bracket color is the same color as the indentation color.

There is a more reasonable workaround for vanilla HTML but it requires you to always close the elements (even void elements).
```json
"[html]": {
    "editor.language.colorizedBracketPairs": [
      ["<", "</"],
      ["<", "/>"]
    ],
```
![image](https://github.com/microsoft/vscode/assets/110549389/956ff002-f89c-4127-9211-6cd0aee04850)

A real fix would be better though, especially for code where HTML is included as markup.

**Comment by developer-256** [comment]
> A real fix would be better though, especially for code where HTML is included as markup.

It isn't a fix. Rather it makes your code much more messed up because of tags like `<img src="" alt="">` or `<input type="text">` that do not have a closing tag and this code snippet matches it with next `/>` and in html both these tags are necessity so entire code becomes very messed up.

For reference look at the picture:
<img width="131" alt="image" src="https://github.com/microsoft/vscode/assets/121548329/175e63d9-ee0c-4a1d-8364-62784b328aa1">

Test this theory with a long html code and you will get the idea.

**Comment by daslicht** [comment]
> > > That should be it.
> > 
> > 
> > doesnt work here, no matter where i click the tag
> 
> <img alt="image" width="87" src="https://user-images.githubusercontent.com/121548329/282301569-51b56643-53d3-41c0-ac03-1952190f5d89.png">
That just highlight the indention and wont work if code is indented like this:

![](https://i.imgur.com/3lGtRGc.png)

**Comment by AntonSvaromir** [comment]
![2024-03-19_20-09-20](https://github.com/microsoft/vscode/assets/130767595/9e2df9f3-5e9d-427c-83ce-44296fd2596d)



``` typescript
"editor.language.colorizedBracketPairs": [
		["<", "/>"],
		["<", ">"],
		["</", ">"],
		["{", "}"],
		["[", "]"],
		["(", ")"]		
	]
``` 
It works fine, only the arrow function looks strange, the arrow is red.

**Comment by fwqaaq** [comment]
> It works fine, only the arrow function looks strange, the arrow is red.

Yeah, how to solve this problem, the red looks very uncomfortable.

**Comment by henrikvilhelmberglund** [comment]
> > It works fine, only the arrow function looks strange, the arrow is red.
> 
> Yeah, how to solve this problem, the red looks very uncomfortable.

Just remove `["<", ">"],`.

**Comment by Ken-SBIT** [comment]
> > > It works fine, only the arrow function looks strange, the arrow is red.
> > 
> > 
> > Yeah, how to solve this problem, the red looks very uncomfortable.
> 
> Just remove `["<", ">"],`.

Nope. That mucks up regular tags. It messes up greater-than comparisons, too.

**Comment by baicaitomato** [comment]
bumppp
